packages/relay/tests/lib/subscriptionController.spec.ts

```diff
@@ -0,0 +1,183 @@
+       });
+       // @ts-ignore
+       ethImpl = new EthImpl(sdkClientStub, mirrorNodeInstance, logger, '0x12a', cache);
+       poller = new Poller(ethImpl, logger);
+
+       subscriptionController = new SubscriptionController(poller, logger);
+   });
+
+   this.beforeEach(()=> {
+       cache.clear();
+       restMock.reset();
+       sandbox = sinon.createSandbox();
+   });
+
+    this.afterEach(() => {
+        restMock.resetHandlers();
+        sandbox.restore();
+    });
+
+    it('Should create sha256 hash out of a data object', async function () {
+        const dataToHash = "This is a Test";
+
+        const hash = subscriptionController.createHash(dataToHash);
+
+        expect(hash).to.be.eq(`401b022b962452749726ba96d436921e39d6deb2b0f4a922cc3da5d7e99e6e46`);
+    });
+
+    it('generateId should create a random hex ID, with 34 length', async function () {
+        const generatedId = subscriptionController.generateId();
+
+        expect(generatedId).to.be.length(34);
+    });
+
+    it('generatedId should be unique', async () => {
+        const generatedId = subscriptionController.generateId();
+        const generatedId2 = subscriptionController.generateId();
+
+        expect(generatedId).not.to.be.eq(generatedId2);
+    });
+
+    it('when subscribing should return subId and poller should add(tag)', async function () {
+        const wsConnection = new MockWsConnection("1");
+        const spy = sandbox.spy(poller, 'add');
+
+        const subId = subscriptionController.subscribe(wsConnection, 'logs');
+
+        expect(spy.args[0][0]).to.be.eq(`{"event":"logs"}`);
+        expect(subId).to.be.length(34);
+    });
+
+    it('notifySubscribers should notify subscribers with data', async function () {
+        const wsConnection = new MockWsConnection("2");
+        const subId = subscriptionController.subscribe(wsConnection, 'logs');
+        const spy = sandbox.spy(wsConnection, 'send');
+        const testData = "test example data";
+
+        subscriptionController.notifySubscribers(`{"event":"logs"}`, testData);
+
+        expect(spy.args[0][0]).to.be.eq(`{"method":"eth_subscription","params":{"result":"${testData}","subscription":"${subId}"}}`);
+    });
+
+    it('notifySubscribers should use cache to not send the data again', async function () {
+        const wsConnection = new MockWsConnection("3");
+        const subId = subscriptionController.subscribe(wsConnection, 'logs');
+        const spy = sandbox.spy(wsConnection, 'send');
+        const testData = "test example data cached";
+
+        subscriptionController.notifySubscribers(`{"event":"logs"}`, testData);
+        subscriptionController.notifySubscribers(`{"event":"logs"}`, testData); // should hit cache
+        subscriptionController.notifySubscribers(`{"event":"logs"}`, testData); // should hit cache
+
+        expect(spy.args[0][0]).to.be.eq(`{"method":"eth_subscription","params":{"result":"${testData}","subscription":"${subId}"}}`);
```

Comment:
nit: this syntax is somewhat more readable:
```
spy.getCall(0).calledWith(`{"method":"eth_subscription","params":{"result":"${testData}","subscription":"${subId}"}}`);
```

Reply:
Thanks for the suggestion, you are right, am fairly new to sinon and nodejs ecosystem in general, so keep the suggestions and tips coming :)

Reply:
I used `expect(spy.getCall(0).args[0]).to.be.eq('{"event":"logs"}');`
as opposed of `assert(spy.getCall(0).calledWith('{"event":"logs"}'));`
Since that way if there is an unexpected value, the assertion just fails, where the first one shows the expected and the actual values, and that makes easier to fix or understand.

Reply:
I've also added some negative test cases ✅ 
and multiple subscribers ✅